core/formatter.cpp

```diff
@@ -1724,16 +1733,233 @@ class FixIndentation {
+     */
+    std::pair<Fodder, Fodder> splitOpenFodder(AST *expr) {
+        const Fodder &fodder = open_fodder(expr);
+        Fodder afterPrev, beforeThis;
+        bool inSecondPart = false;
+        for (const auto &fodderElem: fodder) {
+            if (inSecondPart) {
+                fodder_push_back(beforeThis, fodderElem);
+            } else {
+                afterPrev.push_back(fodderElem);
+            }
+            if (fodderElem.kind != FodderElement::Kind::INTERSTITIAL) {
+                inSecondPart = true;
+                if (fodderElem.blanks > 0) {
+                    afterPrev.back().blanks = 0;
+                    assert(beforeThis.empty());
+                    beforeThis.emplace_back(
+                        FodderElement::Kind::LINE_END,
+                        fodderElem.blanks,
+                        fodderElem.indent,
+                        std::vector<std::string>()
+                    );
+                }
+            }
+        }
+        return {afterPrev, beforeThis};
+    }
+
+    void sortGroup(ImportElems &imports) {
+        // We don't want to change behaviour in such case:
+        // local foo = "b.jsonnet";
+        // local foo = "a.jsonnet";
+        // So we don't change the order when there are shadowed variables.
+        if (!duplicatedVariables(imports)) {
+            std::sort(imports.begin(), imports.end());
+        }
+    }
+
+    ImportElems extractImportElems(Local *local, Fodder after) {
+        ImportElems result;
+        for (auto &bind: local->binds) {
+            Import *import = dynamic_cast<Import*>(bind.body);
+            assert(import != nullptr);
+            result.emplace_back(sortingKey(import), after, &bind);
+            after = {FodderElement(FodderElement::Kind::LINE_END, 0, 0, {})};
+        }
+        return result;
+    }
 
+    AST *buildGroupAST(ImportElems &imports, AST *body, const Fodder &groupOpenFodder) {
+        for (int i = imports.size() - 1; i >= 0; --i) {
+            auto &import = imports[i];
+            Fodder fodder;
+            if (i == 0) {
+                fodder = groupOpenFodder;
+            } else {
+                fodder = imports[i - 1].afterFodder;
+            }
+            Local::Binds binds({*import.bind});
+            auto *local = alloc.make<Local>(
+                LocationRange(),
+                fodder,
+                binds,
+                body
+            );
+            body = local;
+        }
 
-// TODO(dcunnin): Add pass to alphabeticize top level imports.
+        return body;
+    }
 
+    bool duplicatedVariables(const ImportElems &elems) {
+        std::vector<UString> idents;
+        for (const auto &elem: elems) {
+            idents.push_back(elem.bind->var->name);
+        }
+        return std::unique(idents.begin(), idents.end()) != idents.end();
+    }
+
+    /// Check if the import group ends after this local
+    bool groupBoundaryAfter(Local *local) {
```

Comment:
This doesn't modify the input `local` contents, right? The current name kind-of suggests that it does - maybe `groupBoundaryStartsAfter` sounds more like you're asking a question rather than asking it to group a boundary.